export Close() method for subscopes

emmanuel-chukwu opened the pull request:
Add ClosableScope interface to embed Scope and expose a Close() method to allow explicit cleanup of subscopes. This enables better resource management by allowing clients to:

- Clear metrics for unused subscopes

- Remove scopes from registry

- Release associated resources

The method is idempotent and thread-safe, returning an error if the scope is already closed.

emmanuel-chukwu committed:
feat: export Close() method for subscopes

Add Close() method to Scope interface to allow explicit cleanup of subscopes. This enables better resource management by allowing clients to:

- Clear metrics for unused subscopes

- Remove scopes from registry

- Release associated resources

The method is idempotent and thread-safe, returning an error if the scope is already closed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/uber-go/tally?pullRequest=271) <br/>All committers have signed the CLA.

prateek commented on types.go:
```diff
@@ -62,6 +62,12 @@ type Scope interface {
+	// For root scopes, this will also close the reporter.
+	// For subscopes, this will clear all metrics and remove the scope from the registry.
+	// Returns an error if the scope is already closed.
+	Close() error
```
can't do this - it's a breaking change to add a method to a go interface in semver terms

emmanuel-chukwu replied:
I've introduced a new ClosableScope interface to embed Scope with a Close() method

emmanuel-chukwu committed:
Add ClosableScope interface to expose Close() on subscopes in a backwards-compatible way

- Introduce ClosableScope interface embedding Scope with Close() method
- Add NewClosableRootScope, TaggedClosable, and SubScopeClosable methods to obtain subscopes as ClosableScope
- Add unit tests for new ClosableScope functionality
- Preserve semver compatibility by not modifying Scope interface or method signatures
